File: DayTrades/PickViewController.m
//
//  PickViewController.m
//  DayTrades
//
//  Created by Jason Wells on 1/9/15.
//  Copyright (c) 2015 Jason Wells. All rights reserved.
//

#import "PickViewController.h"
#import "DayTrades-Swift.h"

@interface PickViewController ()

@property (weak, nonatomic) IBOutlet UIView *detailsView;
@property (weak, nonatomic) IBOutlet UILabel *detailsLabel;
@property (weak, nonatomic) IBOutlet UILabel *tradeDateLabel;
@property (weak, nonatomic) IBOutlet UIView *securityView;
@property (weak, nonatomic) IBOutlet UILabel *symbolLabel;
@property (weak, nonatomic) IBOutlet UILabel *priceLabel;
@property (weak, nonatomic) IBOutlet UILabel *nameLabel;
@property (weak, nonatomic) IBOutlet UILabel *changeLabel;
@property (weak, nonatomic) IBOutlet UILabel *disclaimerLabel;
@property (weak, nonatomic) IBOutlet UIButton *confirmButton;

@property (strong, nonatomic) NSDateFormatter *dateFormatter;
@property (strong, nonatomic) NSString *dayOfTrade;
@property (strong, nonatomic) Quote *quote;

@end

@implementation PickViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    [self.navigationController.navigationBar setBarStyle:UIBarStyleBlack];
    [self.navigationController.navigationBar setTranslucent:YES];
    [self.navigationController.navigationBar setTintColor:[UIColor whiteColor]];
    [self.view setBackgroundColor:[UIColor colorWithPatternImage:[UIImage imageNamed:@"background-2.jpg"]]];
    [self.detailsView setBackgroundColor:[UIColor translucentColor]];
    [self.securityView setBackgroundColor:[UIColor translucentColor]];
    
    self.dateFormatter = [[NSDateFormatter alloc] init];
    self.dayOfTrade = [MarketHelper nextDayOfTrade];
    [self refreshViews];
}

- (BOOL)searchBar:(UISearchBar *)searchBar shouldChangeTextInRange:(NSRange)range replacementText:(NSString *)text {
    return [searchBar.text length] + [text length] - range.length <= 5;
}

- (void)searchBar:(UISearchBar *)searchBar textDidChange:(NSString *)searchText {
    if (![searchText hasPrefix:@"^"]) {
        NSSet *symbols = [[[NSSet alloc] init] setByAddingObject:[searchText uppercaseString]];
        [FinanceClient fetchQuotesForSymbols:symbols block:^(NSURLResponse *response, NSData *data, NSError *error) {
            self.quote = nil;
            if (!error) {
                NSArray *quotes = [Quote fromData:data];
                if (quotes.count == 1) {
                    Quote *quote = [quotes firstObject];
                    if ([quote.symbol isEqualToString:[searchBar.text uppercaseString]]) {
                        self.quote = quote;
                    }
                }
            } else {
                NSLog(@"Error: %@ %@", error, [error userInfo]);
            }
            [self refreshViews];
        }];
    }
}

- (void)refreshViews {
    static NSString *text = @"The listed security will be purchased for the full value of your account at the opening price and sold at market close on %@.";
    NSString *dateFormat = [self.dateFormatter fullFromDayOfTrade:self.dayOfTrade];
    if (self.quote) {
        NSString *disclaimer = [NSString stringWithFormat:text, dateFormat];
        [self.disclaimerLabel setText:disclaimer];
        [self.symbolLabel setText:self.quote.symbol];
        [self.nameLabel setText:self.quote.name];
        [self.priceLabel setText:[NSString stringWithFormat:@"%0.2f", self.quote.price]];
        [self.changeLabel setText:[ChangeFormatter stringFromQuote:self.quote]];
        [self.changeLabel setTextColor:[UIColor changeColor:self.quote.priceChange]];
        [self.detailsView setHidden:YES];
        [self.securityView setHidden:NO];
    }
    else {
        [self.detailsLabel setText:@"Choose a security to buy on"];
        [self.tradeDateLabel setText:dateFormat];
        [self.detailsView setHidden:NO];
        [self.securityView setHidden:YES];
    }
}

- (void)searchBarSearchButtonClicked:(UISearchBar *)searchBar {
    [self.view endEditing:YES];
}

- (IBAction)onConfirmButtonTouched:(id)sender {
    if (self.quote) {
        NSString *symbol = self.quote.symbol;
        NSString *dayOfTrade = [MarketHelper nextDayOfTrade];
        Pick *pick = [Pick newForAccount:self.account symbol:symbol dayOfTrade:dayOfTrade];
        [pick saveInBackgroundWithBlock:^(BOOL succeeded, NSError *error) {
            if (succeeded) {
                [self.delegate updateNextPick:pick];
                [self.navigationController popViewControllerAnimated:YES];
            }
        }];
    }
}

- (IBAction)onTap:(id)sender {
    [self.view endEditing:YES];
}

@end

File: DayTrades/PickViewController.swift
//
//  PickViewController.swift
//  DayTrades
//
//  Created by Jason Wells on 7/12/15.
//  Copyright (c) 2015 Jason Wells. All rights reserved.
//

import Foundation

class PickViewController: UIViewController, UISearchBarDelegate {
    
    @IBOutlet weak var detailsView: UIView!
    @IBOutlet weak var securityView: UIView!
    @IBOutlet weak var detailsLabel: UILabel!
    @IBOutlet weak var dayOfTradeLabel: UILabel!
    @IBOutlet weak var symbolLabel: UILabel!
    @IBOutlet weak var priceLabel: UILabel!
    @IBOutlet weak var nameLabel: UILabel!
    @IBOutlet weak var disclaimerLabel: UILabel!
    @IBOutlet weak var submitButton: UIButton!
    
    let disabledSymbols: NSArray = NSBundle.mainBundle().objectForInfoDictionaryKey("Disabled symbols") as! NSArray
    let dateFormatter: NSDateFormatter = NSDateFormatter()
    let numberFormatter: NSNumberFormatter = NSNumberFormatter()
    
    var quote: Quote?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        if let backgroundImage: UIImage = UIImage(named: "background-2.jpg") {
            view.backgroundColor = UIColor(patternImage: backgroundImage)
        }
        detailsView.backgroundColor = UIColor.translucentColor()
        securityView.backgroundColor = UIColor.translucentColor()
    }
    
    override func viewWillAppear(animated: Bool) {
        super.viewWillAppear(animated)
        refreshView()
    }
    
    func refreshView() {
        let dayOfTrade: String = MarketHelper.nextDayOfTrade()
        let text: String? = dateFormatter.fullTextFromDayOfTrade(dayOfTrade)
        if let quote: Quote = quote {
            disclaimerLabel.text = "The listed security will be purchased for the full value of your account at the opening price and sold at market close on \(text!)."
            symbolLabel.text = quote.symbol
            nameLabel.text = quote.name
            priceLabel.text = numberFormatter.priceFromNumber(NSNumber(double: quote.price))
            detailsView.hidden = true
            securityView.hidden = false
        }
        else {
            detailsLabel.text = "Choose a security to buy on"
            dayOfTradeLabel.text = text!
            securityView.hidden = true
            detailsView.hidden = false
        }
    }
    
    func searchBar(searchBar: UISearchBar, shouldChangeTextInRange range: NSRange, replacementText text: String) -> Bool {
        return count(searchBar.text) + count(text) - range.length <= 5;
    }
    
    func searchBar(searchBar: UISearchBar, textDidChange searchText: String) {
        if isValidSymbol(searchText) {
            let symbols: Set<String> = ["\(searchText.uppercaseString)"]
            FinanceClient.fetchQuotesForSymbols(symbols, block: { (response: NSURLResponse!, data: NSData!, error: NSError!) -> Void in
                self.quote = nil
                if let data: NSData = data {
                    let quotes: Array<Quote> = Quote.fromData(data)
                    if let quote = quotes.first {
                        if quote.symbol == searchBar.text.uppercaseString {
                            self.quote = quote
                        }
                    }
                }
                if let error: NSError = error {
                    println("Error \(error) \(error.userInfo)")
                }
                self.refreshView()
            })
        }
    }
    
    func searchBarSearchButtonClicked(searchBar: UISearchBar) {
        view.endEditing(true)
    }
    
    func isValidSymbol(text: String) -> Bool {
        if text.hasPrefix("^") {
            return false
        }
        if disabledSymbols.containsObject(text.uppercaseString) {
            return false
        }
        return true
    }
    
    @IBAction func onSubmitButtonTouched(sender: AnyObject) {
        if let quote: Quote = self.quote {
            if let symbol: String = quote.symbol {
                submitButton.enabled = false
                ParseClient.setNextPick(symbol, block: { (object: PFObject?, error: NSError?) -> Void in
                    if let nextPick: Pick = object as? Pick {
                        NSNotificationCenter.defaultCenter().postNotificationName(Notification.NextPickUpdated.description, object: nil)
                        self.submitButton.enabled = true
                        self.navigationController?.popViewControllerAnimated(true)
                    }
                })
            }
        }
    }

    @IBAction func onTap(sender: AnyObject) {
        view.endEditing(true)
    }
    
}



